addon/state-for.js

```diff
@@ -22,8 +23,26 @@ function createStates(container, stateName) {
     throw new TypeError(`Unknown StateFactory: \`${stateContainerName}\``);
   }
 
-  states[stateName] = new Ember.MapWithDefault({
-    defaultValue: () => buildDefaultState.call(this, StateFactory)
+  /*
+  * Overriding the default get function to support the defaultValue function
+  * recieving a custom context.
+  * Taken from:
+  * https://github.com/emberjs/ember.js/blob/6a9c789295/packages/ember-metal/lib/map.js#L467
+  */
+  MapWithDefault.prototype.get = function(key, context) {
```

Comment:
a simple bind wont work in this case:

```js

// components/foo.js
export default Ember.Component.extend({
  data: stateFor('email', {key: 'somekey'})
});


//components/bar.js
export default Ember.Component.extend({
  data: stateFor('email', {key: 'somekey'})
});
```

If 2 components use the same state then each component's stateFor calls:

```js
states[stateName].get(this.get(key))
```

However, defaultValues context will be 'binded' to the first component and will not be 'dynamic'

```js

new MapWithDefault({
  defaultValue: function() { // how do we make the context here dynamic to each component }
});
```